### PR TITLE
Riak - More statistics

### PR DESCRIPTION
**Pull request by bostko**

### CODE REVIEW

File: software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java [hunk 1]
```diff
@@ -142,6 +140,27 @@
...
+    AttributeSensor<String> SEARCH_INDEX_LATENCY_MEDIAN = Sensors.newStringSensor("search_index_latency_median", "Time it takes Search to index a new document");
+    AttributeSensor<String> CONSISTENT_GET_TIME_MEAN = Sensors.newStringSensor("consistent_get_time_mean", "Strongly consistent read latency");
+    AttributeSensor<String> CONSISTENT_PUT_TIME_MEAN = Sensors.newStringSensor("consistent_put_time_mean", "Strongly consistent write latency");
+
+    List<AttributeSensor<String>> ONE_MINUTE_STRING_SENSORS = ImmutableList.of(NODE_GET_FSM_TIME_MEAN, NODE_PUT_FSM_TIME_MEAN,
+            OBJECT_COUNTER_MERGE_TIME_MEAN, OBJECT_SET_MERGE_TIME_MEAN, OBJECT_MAP_MERGE_TIME_MEAN,
+            SEARCH_QUERY_LATENCY_MEDIAN, SEARCH_INDEX_LATENCY_MEDIAN,
+            CONSISTENT_GET_TIME_MEAN, CONSISTENT_PUT_TIME_MEAN);
+
+    ConfigKey<Boolean> DO_OS_TUNING = new BasicConfigKey<Boolean>(
```
Comment (grkvlt) [inline software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java]: Use `ConfigKeys.newBooleanConfigKey` here, although this seems to be a duplicate of https://github.com/apache/incubator-brooklyn/blob/master/software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java#L49-L50 from #563

File: software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java [hunk 2]
```diff
@@ -142,6 +140,27 @@
...
+            Boolean.class,
+            "do.os.tunning",
```
Comment (grkvlt) [inline software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java]: Also, this is not a good configuration key name, they should be prefixed `riak.` for this entity

File: software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNodeImpl.java [hunk 3]
```diff
@@ -161,8 +162,15 @@ public void connectSensors() {
                                     }
                                 }
                         ))
-                        .onFailureOrException(Functions.constant(Arrays.asList(new String[0]))))
-                .build();
+                        .onFailureOrException(Functions.constant(Arrays.asList(new String[0]))));
+
+        for(AttributeSensor<String> sensor : ONE_MINUTE_STRING_SENSORS) {
```
Comment (grkvlt) [inline software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNodeImpl.java]: Should be a space after `for` and before `(`

File: software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNodeImpl.java [hunk 4]
```diff
@@ -161,8 +162,15 @@ public void connectSensors() {
...
+            httpFeedBuilder.poll(new HttpPollConfig<String>(sensor)
+                    .onSuccess(HttpValueFunctions.jsonContents(sensor.getName(), String.class))
+                    .onFailureOrException(Functions.constant("-1")));
```
Comment (grkvlt) [inline software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNodeImpl.java]: Why are we returning a string with a number in it? If these are numeric sensors, then parse the results with `Integer#valueOf(String)` or similar, so they can be used in policies.

File: software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java [hunk 5]
```diff
@@ -142,6 +140,27 @@
     MethodEffector<Void> LEAVE_RIAK_CLUSTER = new MethodEffector<Void>(RiakNode.class, "leaveCluster");
     MethodEffector<Void> COMMIT_RIAK_CLUSTER = new MethodEffector<Void>(RiakNode.class, "commitCluster");
 
+    AttributeSensor<String> NODE_GET_FSM_TIME_MEAN = Sensors.newStringSensor("node_get_fsm_time_mean", "Time between reception of client read request and subsequent response to client");
```
Comment (grkvlt) [inline software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java]: Sensor names should be of the form `riak.node.fsm_time.mean` and should indicate in the description what the units are (milliseconds, perhaps?)

File: software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java [hunk 6]
```diff
@@ -142,6 +140,27 @@
...
+    AttributeSensor<String> NODE_PUT_FSM_TIME_MEAN = Sensors.newStringSensor("node_put_fsm_time_mean", "Time between reception of client write request and subsequent response to client");
+    AttributeSensor<String> OBJECT_COUNTER_MERGE_TIME_MEAN = Sensors.newStringSensor("object_counter_merge_time_mean", "Time it takes to perform an Update Counter operation");
+    AttributeSensor<String> OBJECT_SET_MERGE_TIME_MEAN = Sensors.newStringSensor("object_set_merge_time_mean", "Time it takes to perform an Update Set operation");
+    AttributeSensor<String> OBJECT_MAP_MERGE_TIME_MEAN = Sensors.newStringSensor("object_map_merge_time_mean", "Time it takes to perform an Update Map operation");
+    AttributeSensor<String> SEARCH_QUERY_LATENCY_MEDIAN = Sensors.newStringSensor("search_query_latency_median", "Search query latency");
```
Comment (grkvlt) [inline software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java]: This one should be `riak.node.searchQueryLatency.median` for example

File: software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java [hunk 7]
```diff
@@ -142,6 +140,27 @@
     MethodEffector<Void> LEAVE_RIAK_CLUSTER = new MethodEffector<Void>(RiakNode.class, "leaveCluster");
     MethodEffector<Void> COMMIT_RIAK_CLUSTER = new MethodEffector<Void>(RiakNode.class, "commitCluster");
 
+    AttributeSensor<String> NODE_GET_FSM_TIME_MEAN = Sensors.newStringSensor("node_get_fsm_time_mean", "Time between reception of client read request and subsequent response to client");
+    AttributeSensor<String> NODE_PUT_FSM_TIME_MEAN = Sensors.newStringSensor("node_put_fsm_time_mean", "Time between reception of client write request and subsequent response to client");
```
Comment (grkvlt) [inline software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java]: This one should be  `riak.node.fsm_time.put.mean`

File: software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java [hunk 8]
```diff
@@ -142,6 +140,27 @@
...
+    AttributeSensor<String> OBJECT_COUNTER_MERGE_TIME_MEAN = Sensors.newStringSensor("object_counter_merge_time_mean", "Time it takes to perform an Update Counter operation");
+    AttributeSensor<String> OBJECT_SET_MERGE_TIME_MEAN = Sensors.newStringSensor("object_set_merge_time_mean", "Time it takes to perform an Update Set operation");
```
Comment (grkvlt) [inline software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java]: This one should be `riak.node.merge_time.object_set.mean`

File: software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java [hunk 9]
```diff
@@ -142,6 +140,27 @@
...
+    AttributeSensor<String> OBJECT_MAP_MERGE_TIME_MEAN = Sensors.newStringSensor("object_map_merge_time_mean", "Time it takes to perform an Update Map operation");
```
Comment (grkvlt) [inline software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java]: This one should be `riak.node.merge_time.object_map.mean`

File: software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java [hunk 10]
```diff
@@ -142,6 +140,27 @@
     MethodEffector<Void> LEAVE_RIAK_CLUSTER = new MethodEffector<Void>(RiakNode.class, "leaveCluster");
     MethodEffector<Void> COMMIT_RIAK_CLUSTER = new MethodEffector<Void>(RiakNode.class, "commitCluster");
 
+    AttributeSensor<String> NODE_GET_FSM_TIME_MEAN = Sensors.newStringSensor("node_get_fsm_time_mean", "Time between reception of client read request and subsequent response to client");
+    AttributeSensor<String> NODE_PUT_FSM_TIME_MEAN = Sensors.newStringSensor("node_put_fsm_time_mean", "Time between reception of client write request and subsequent response to client");
+    AttributeSensor<String> OBJECT_COUNTER_MERGE_TIME_MEAN = Sensors.newStringSensor("object_counter_merge_time_mean", "Time it takes to perform an Update Counter operation");
```
Comment (grkvlt) [inline software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java]: This one should be `riak.node.merge_time.object_counter.mean`

File: software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java [hunk 11]
```diff
@@ -142,6 +140,27 @@
...
+    AttributeSensor<String> OBJECT_SET_MERGE_TIME_MEAN = Sensors.newStringSensor("object_set_merge_time_mean", "Time it takes to perform an Update Set operation");
+    AttributeSensor<String> OBJECT_MAP_MERGE_TIME_MEAN = Sensors.newStringSensor("object_map_merge_time_mean", "Time it takes to perform an Update Map operation");
+    AttributeSensor<String> SEARCH_QUERY_LATENCY_MEDIAN = Sensors.newStringSensor("search_query_latency_median", "Search query latency");
+    AttributeSensor<String> SEARCH_INDEX_LATENCY_MEDIAN = Sensors.newStringSensor("search_index_latency_median", "Time it takes Search to index a new document");
```
Comment (grkvlt) [inline software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java]: This one should be `riak.node.latency.search_index.median`

File: software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java [hunk 12]
```diff
@@ -142,6 +140,27 @@
...
+    AttributeSensor<String> CONSISTENT_GET_TIME_MEAN = Sensors.newStringSensor("consistent_get_time_mean", "Strongly consistent read latency");
```
Comment (grkvlt) [inline software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java]: This one should be `riak.node.time.consistent_get.mean`

File: software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java [hunk 13]
```diff
@@ -142,6 +140,27 @@
...
+    AttributeSensor<String> CONSISTENT_PUT_TIME_MEAN = Sensors.newStringSensor("consistent_put_time_mean", "Strongly consistent write latency");
```
Comment (grkvlt) [inline software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java]: This one should be `riak.node.time.consistent_put.mean`

**Comment by grkvlt** [comment]
I understand why the sensors have the names they were given, but it would still be possible to generate the JSON parsing code in the `HttpFeed` if they had a `riak.node.` prefix. You should also be able to either parse or cast the reult to an integer representing the time.

**Comment by bostko** [comment]
@grkvlt I put only `riak.` prefix since several of the parameters already has the word node in their names.
Is that OK?

**Comment by grkvlt** [comment]
@bostko yes, that looks good now, am happy to merge now

File: software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java [hunk 14]
```diff
@@ -29,12 +29,10 @@
 import brooklyn.entity.basic.SoftwareProcess;
 import brooklyn.entity.proxying.ImplementedBy;
 import brooklyn.event.AttributeSensor;
-import brooklyn.event.basic.AttributeSensorAndConfigKey;
-import brooklyn.event.basic.PortAttributeSensorAndConfigKey;
-import brooklyn.event.basic.Sensors;
-import brooklyn.event.basic.TemplatedStringAttributeSensorAndConfigKey;
+import brooklyn.event.basic.*;
```
Comment (grkvlt) [inline software/nosql/src/main/java/brooklyn/entity/nosql/riak/RiakNode.java]: No wildcard imports allowed. The standard for Brooklyn Java code is essentially the original Sun **Java Code Conventions** (from 1997) which you can also find at Oracle [here](http://www.oracle.com/technetwork/java/codeconventions-150003.pdf).

**Comment by grkvlt** [comment]
So, if you fix the imports I'm happy merging this now.

**Comment by grkvlt** [comment]
Build failure is due to lack of license in files added by 89b621b1b2eeabc631cea77f10e9ad626ee072d6 so can be ignored - the tests are still passing in the `software/nosql` project

**Comment by bostko** [comment]
Wait for me to test it. There is a problem related with renaming the config parameters.

**Pull request merged**